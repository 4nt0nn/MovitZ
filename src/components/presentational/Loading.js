import React from "react";

/**
 * Functional component for rendering
 * a loading screen with dynamic loading text.
 *
 * @param {object} props - containing the loading text.
 */
const Loading = (props) => {
  return (
    <div className={"loading-screen valign-wrapper"}>
      <h5 className={"white-text"}>{props.text}</h5>
      <div className={"progress container"}>
        <div className={"indeterminate"}></div>
      </div>
    </div>
  );
};

export default Loading;
