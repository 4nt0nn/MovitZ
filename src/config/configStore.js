import { applyMiddleware, createStore } from "redux";
import thunkMiddleware from "redux-thunk";
import { composeWithDevTools } from "redux-devtools-extension";
import { getFirebase } from "react-redux-firebase";
import { getFirestore } from "redux-firestore";

import rootReducer from "../store/reducers/root";

/**
 * Configuration function for handling all
 * configurations to the createStore method.
 *
 * Split from index.js for better maintainability.
 *
 * Applying thunk as middleware in order to halt
 * the dispatch from action creators which will
 * let us fetch data and dispatch actions in callbacks.
 *
 * @param {object} preloadedState - Added in case we need to add
 * a preloaded state later.
 */
export default function configureStore(preloadedState) {
  /**
   * Passing extra arguments to middleware in order to use getFirebase/getFirestore
   * in our action creators to communicate with firebase/firestore.
   */
  const middlewares = [
    thunkMiddleware.withExtraArgument({ getFirebase, getFirestore }),
  ];
  const middlewareEnhancer = applyMiddleware(...middlewares);

  const enhancers = [middlewareEnhancer];
  const composedEnhancers = composeWithDevTools(...enhancers);

  const store = createStore(rootReducer, preloadedState, composedEnhancers);

  return store;
}
