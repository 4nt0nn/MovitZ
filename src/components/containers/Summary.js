import React, { useEffect } from "react";
import { useParams } from "react-router-dom";
import { useDispatch, useSelector } from "react-redux";
import { useFirestore } from "react-redux-firebase";
import { Link } from "react-router-dom";

import { tryAddMovie, tryGetMovie } from "../../store/actions/movie";
import imagePlaceholder from "../../images/missing.png";
import Loading from "../presentational/Loading";
import Card from "../presentational/Card";

/**
 * Functional component for rendering
 * the movie summary page of the application.
 *
 * Handles the dispatch of action creators
 * for fetching a specific movie
 * from TMDBs API
 *
 * Handles adding movies to the firestore database.
 *
 */
const Summary = () => {
  const dispatch = useDispatch();
  const firestore = useFirestore();
  const { id } = useParams();
  const movie = useSelector((state) => state.tmdb.movie);
  const auth = useSelector((state) => state.firebase.auth);

  const boundAddMovies = (firestore, auth, movie, favourite) =>
    dispatch(tryAddMovie(firestore, auth, movie, favourite));

  useEffect(() => {
    dispatch(tryGetMovie(id));
  }, [dispatch, id]);

  const onAddToMoviesHandler = (movie, favourite) => {
    boundAddMovies(firestore, auth, movie, favourite);
  };

  if (movie) {
    const { backdrop_path, title, overview, genres } = movie;
    const movieGenres = genres.map((genre) => (
      <p key={genre.id} className={"yellow-text darken-3"}>
        {genre.name}
      </p>
    ));
    return (
      <div className={"row summary-container"}>
        <div className={"summary-image-container"}>
          <Link
            to={"/"}
            className={
              "btn-floating btn-small waves-effect waves-light grey darken-3 summary-back-link"
            }
          >
            <i className={"material-icons"}>keyboard_arrow_left</i>
          </Link>
          <img
            className={"movie-backdrop"}
            src={
              backdrop_path != null
                ? `https://image.tmdb.org/t/p/original${backdrop_path}`
                : imagePlaceholder
            }
            alt={`${title}'s poster`}
          />
          <div className={"summary-movie-title-card-overview"}>
            <h5 className={"white-text summary-movie-title"}>{title}</h5>

            <div className={"summary-card"}>
              <Card
                addToWatchLater={(movie) => onAddToMoviesHandler(movie, false)}
                addToFavourites={(movie) => onAddToMoviesHandler(movie, true)}
                displayTitle={false}
                card={movie}
              />
            </div>
            <div className={"summary-overview-genres"}>
              <p className={"grey-text lighten-4 summary-movie-overview"}>
                {overview ? overview : "This movie has not summary"}
              </p>
              {movieGenres}
            </div>
          </div>
        </div>
      </div>
    );
  } else {
    return <Loading text={"Loading content..."} />;
  }
};

export default Summary;
