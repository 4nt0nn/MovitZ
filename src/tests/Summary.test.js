import React from "react";
import { shallow, mount } from "enzyme";
import { MemoryRouter } from "react-router-dom";
import Summary from "../components/containers/Summary";
import Loading from "../components/presentational/Loading";
import { Provider } from "react-redux";
import configStore from "../config/configStore";

describe("<Summary />", () => {
  it('should render correctly in "debug" mode', () => {
    const wrapper = shallow(
      <Provider store={configStore()}>
        <Summary debug />
      </Provider>
    );
    expect(wrapper).toMatchSnapshot();
  });
  it('should render loading since auth is not loaded in "debug" mode', () => {
    const movie = {
      title: "Star Wars",
      backdrop_path: "/test",
      overview: "test",
      genres: [],
    };
    const wrapper = mount(
      <Provider store={configStore()}>
        <MemoryRouter>
          <Summary movie={movie} debug />
        </MemoryRouter>
      </Provider>
    );
    expect(wrapper.containsMatchingElement(<Loading />)).toBeTruthy();
    wrapper.unmount();
  });
});
