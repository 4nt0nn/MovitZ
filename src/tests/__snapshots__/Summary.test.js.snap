// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Summary /> should render correctly in "debug" mode 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Summary
    debug={true}
  />
</ContextProvider>
`;

exports[`<Summary /> should render with child components in "debug" mode 1`] = `
<Provider
  store={
    Object {
      "dispatch": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
      Symbol(observable): [Function],
    }
  }
>
  <MemoryRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "t59vpt",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "t59vpt",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Summary
        debug={true}
        movie={
          Object {
            "title": "Star Wars",
          }
        }
      >
        <Loading
          text="Loading content..."
        >
          <div
            className="loading-screen valign-wrapper"
          >
            <h5
              className="white-text"
            >
              Loading content...
            </h5>
            <div
              className="progress container"
            >
              <div
                className="indeterminate"
              />
            </div>
          </div>
        </Loading>
      </Summary>
    </Router>
  </MemoryRouter>
</Provider>
`;

exports[`Summary should render correctly in "debug" mode 1`] = `
<ContextProvider
  value={
    Object {
      "store": Object {
        "dispatch": [Function],
        "getState": [Function],
        "replaceReducer": [Function],
        "subscribe": [Function],
        Symbol(observable): [Function],
      },
      "subscription": Subscription {
        "handleChangeWrapper": [Function],
        "listeners": Object {
          "notify": [Function],
        },
        "onStateChange": [Function],
        "parentSub": undefined,
        "store": Object {
          "dispatch": [Function],
          "getState": [Function],
          "replaceReducer": [Function],
          "subscribe": [Function],
          Symbol(observable): [Function],
        },
        "unsubscribe": null,
      },
    }
  }
>
  <Summary
    debug={true}
  />
</ContextProvider>
`;
